src/Explorer.js

```diff
@@ -1096,113 +1195,142 @@ function memoizeParseQuery(query: string): DocumentNode {
+        selectionSet: {
+          ...operationDef.selectionSet,
+          selections,
+        },
+      };
+    } else if (operationDef.kind === 'OperationDefinition') {
+      newOperationDef = {
         ...operationDef,
         selectionSet: {
           ...operationDef.selectionSet,
           selections,
         },
       };
-      let replaced = false;
-      const newDefinitions = parsedQuery.definitions.map(op => {
-        if (op === operationDef) {
-          replaced = true;
-          return newOperationDef;
-        } else {
-          return op;
-        }
-      });
-
-      this.props.onEdit(
-        print({
-          ...parsedQuery,
-          definitions: replaced
-            ? newDefinitions
-            : [newOperationDef, ...newDefinitions],
-        }),
-      );
     }
+
+    this.props.onEdit(newOperationDef);
   };
 
   render() {
     const {
       fields,
       operation,
-      parsedQuery,
+      name,
+      definition,
       schema,
       getDefaultFieldNames,
     } = this.props;
-    const operationDef = this._getOperationDef(parsedQuery);
+    const operationDef = definition;
     const selections = operationDef.selectionSet.selections;
+
+    const operationDisplayName = this.props.name || `Name your ${operation}`;
+
     return (
       <div
+        id={`${operation}-${name || 'unknown'}`}
         style={{
           borderBottom: operation !== 'subscription' ? '1px solid #d6d6d6' : '',
           marginBottom: '1em',
           paddingBottom: '0.5em',
         }}>
-        <div style={{color: '#B11A04', paddingBottom: 4}}>{operation}</div>
-        {Object.keys(fields)
+        <div style={{color: '#B11A04', paddingBottom: 4}}>
+          {operation}{' '}
+          <span style={{color: 'rgb(193, 42,80)'}}>
+            <input
+              style={{
+                border: 'none',
+                borderBottom: '1px solid #888',
+                outline: 'none',
+                color: 'rgb(193, 42,80)',
+                width: `${Math.max(4, operationDisplayName.length)}ch`,
+              }}
+              autoComplete="false"
+              placeholder={`Name your ${operation}`}
+              value={this.props.name}
+              onKeyDown={event => {
+                if (isRunShortcut(event)) {
+                  this.props.onRunOperation(this.props.name);
+                }
+              }}
+              onChange={event =>
+                this.props.onOperationRename(event.target.value)
+              }
+            />
+          </span>
+          {!!this.props.onTypeName ? (
+            <span>
+              <br />
+              {`on ${this.props.onTypeName}`}
+            </span>
+          ) : (
+            ''
+          )}
+        </div>
+
+        {Object.keys(fields || {})
           .sort()
           .map(fieldName => (
             <FieldView
               key={fieldName}
-              field={fields[fieldName]}
+              field={(fields || {})[fieldName]}
```

Comment:
Do you need `|| {}` here? Shouldn't be possible to get here with null fields.

Reply:
Done

Reply:
Flow will be unhappy since the type is `?GraphQLFieldMap<any, any>`, and it's pretty tough to get Flow to agree it's non-nullable (`queryFields/mutationFields/subscriptionFields/fragmentFields` could all be null)